### PR TITLE
lxd/cgroup: Initial abstraction for cgroupsv2 with pids support

### PR DESCRIPTION
**Pull request by shreyassood**
Initial abstraction to support different properties and controllers for https://github.com/lxc/lxd/issues/6305.

### CODE REVIEW

File: lxd/cgroup/cgroup.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
+package cgroup
+
+import (
+	"fmt"
+	"gopkg.in/lxc/go-lxc.v2"
```
Comment (brauner) [inline lxd/cgroup/cgroup.go]: I think it's customary that we leave a \n between native go imports and our imports.

File: lxd/cgroup/cgroup.go [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+	configs, e := SetConfigMap(property, value)
+	if e != nil {
+		return e
+	}
+
+	for _, rule :=  range configs {
+		err := c.SetCgroupItem(rule.Key, rule.Value)
+
+		if err != nil {
```
Comment (brauner) [inline lxd/cgroup/cgroup.go]: No \n between assignment and error checking.

File: lxd/container_lxc.go [hunk 3]
```diff
@@ -3769,7 +3769,28 @@ func (c *containerLXC) Rename(newName string) error {
...
+	}
+
+	// Make sure the container is running
+	if !c.IsRunning() {
+		return "", fmt.Errorf("Can't get cgroups on a stopped container")
+	}
+
+	value, err := cgroup.Get(c.c, property)
+
+	if err != nil {
```
Comment (brauner) [inline lxd/container_lxc.go]: No \n between assignment and error handling.

**Comment by brauner** [comment]
> The overall design looks ok to me.
> @stgraber any objections?

Ok, after chatting to @stgraber for a bit we need to redesign this since we have a requirement to have this work equally for containers, vms, and projects so we need something more complex.

**Comment by stgraber** [comment]
I'm working on a short example of the structure I think we should be using for that. It shouldn't be much work to then reshuffle the bits you have on top of it and port the rest of the current cgroup calls over too.

**Comment by shreyassood** [comment]
Thanks! We also created a Draft PR (https://github.com/lxc/lxd/pull/6580) for how we were thinking of doing the OS struct and an example of using that in `cgroup.Get` -- it's not finalized, but your early thoughts on that would be great too! @stgraber

**Comment by stgraber** [comment]
https://github.com/lxc/lxd/pull/6581

**Comment by stgraber** [comment]
That PR effectively adds the structure for the abstraction, including import time logic for detecting all the controllers on V1 and V2 and a ReadWriter implementation for containers which supports both the LXC config file and live cgroup changes.

I've ported `pids.max` over to it as a test and it appears to be working fine here.

**Comment by stgraber** [comment]
With that in place, you'd now want to iterate through all the current uses of cgroups and port all of them to dedicated functions inside `abstraction.go`, then changing the container code to use the abstraction.

This already handles the startup time logic through the `init()` function, so the main missing logic there is for extra features like swap handling (which can directly be done in the function setting memory limits) and we'll likely want an exported function from the `cgroup` package which can be called by `daemon.go` to log status on startup. Effectively a way to log whether we're running in V1-only, V2-only or hybrid and list the active controllers.

**Comment by shreyassood** [comment]
Thanks @stgraber! That looks great. Should we call the init in daemon.go or replace the existing CGroupInit within sys (https://github.com/shreyassood/lxd/blob/cg_start/lxd/cgroup.go) that is initialized in daemon with `os.Init()`. Also, any suggestions on what do we do with certain controller flags such as https://github.com/shreyassood/lxd/commit/1ad30dd72ece0ae7838029e0b72dd406954e163b which won't necessarily work easily with the new cgControllers config?

**Comment by stgraber** [comment]
@shreyassood init functions are automatically called on import, not need to call it.

**Comment by stgraber** [comment]
@shreyassood for the blkio weight file and for the memsw file. We could use the same approach we have right now and expand the logic in `init()` to perform the extra check when hitting that controller on v1, if the feature is detected, a fake `blkio/weight` or `memory/memsw` controller would be added to `cgControllers` pointing to V1. This then makes it easy to check for in the functions in `abstraction.go`.

**Comment by shreyassood** [comment]
Closing because of #6581

**Pull request closed without merging**